integration/ruler_test.go

```diff
@@ -578,6 +578,12 @@ func TestRulerMetricsForInvalidQueriesAndNoFetchedSeries(t *testing.T) {
 			// Very low limit so that ruler hits it.
 			"-querier.max-fetched-chunks-per-query": "5",
 
+			// Do not involve the block storage as we don't upload blocks and
+			// it can happen that a query runs faster than trying to load the
+			// index, which results in context canceled, which would be cached
+			// in the index loader, failing all subsequent queries.
```

Comment:
this sounds like a bug 😄

Reply:
In non-test use, the context used for loading the index-header is the context associated with the top-level store-gateway `Service` (passed through many many method calls) so this isn't an issue in practice. Guess it still wouldn't hurt to _not_ consider context cancellation a permanent error in the lazy loading of the index-header.

Reply:
i assumed this test was talking about the bucket index loader https://github.com/grafana/mimir/blob/ff8a70aee6d187113d0f2b18b75144a4bee5091d/pkg/querier/blocks_finder_bucket_index.go#L46

I think this is the bug @krajorama referred to 

https://github.com/grafana/mimir/blob/8f53869d0cc45a2773a59a03d43979b963069da1/pkg/storage/tsdb/bucketindex/loader.go#L109-L113

Reply:
I'd say it's more a theoretical bug that happened to happen in this particular integration test. I have no idea if it could happen in real life that one part of the query finishes so fast that in cancels loading stuff from store-gateways which leads to caching context canceled error.

Reply:
the query can be cancelled by the user too. I opened #7620 with a fix. It's a simple change